### PR TITLE
Updates DTLSv1_listen to support DTLS 1.3 client hellos.

### PR DESCRIPTION
**Pull request by fwh-dc**
##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

File: ssl/d1_lib.c [hunk 1]
```diff
@@ -400,13 +401,172 @@ int dtls1_handle_timeout(SSL_CONNECTION *s)
...
+     */
+    int version = (ssl->method->version == DTLS_ANY_VERSION) ? DTLS1_VERSION
+                                                         : s->version;
+
+    if (!WPACKET_init_static_len(wpkt,
+                                 wbuf,
+                                 ssl_get_max_send_fragment(s)
+                                 + DTLS1_RT_HEADER_LENGTH,
+                                 0)
+        || !WPACKET_put_bytes_u8(wpkt, SSL3_RT_HANDSHAKE)
```
Comment (fwh-dc) [inline ssl/d1_lib.c:432]: Fix indentation

File: ssl/d1_lib.c [hunk 2]
```diff
@@ -608,12 +777,54 @@ int DTLSv1_listen(SSL *ssl, BIO_ADDR *client)
             goto end;
         }
 
+        // Check if this is a DTLSv1.3 record
+        int is_dtls13_record = PACKET_remaining(&msgpayload) > 0
+                               && clientvers == DTLS1_2_VERSION;
+
+        int server_accepts_dtls13 = ssl->method->version == DTLS_ANY_VERSION
```
Comment (fwh-dc) [inline ssl/d1_lib.c:784]: Move declarations

**Comment by t8m** [comment]
The CI is relevant.

File: ssl/d1_lib.c [hunk 3]
```diff
@@ -656,16 +923,15 @@ int DTLSv1_listen(SSL *ssl, BIO_ADDR *client)
...
             } else {
                 /* Cookie verification succeeded */
                 next = LISTEN_SUCCESS;
             }
         }
 
-        if (next == LISTEN_SEND_VERIFY_REQUEST) {
+        if (next == LISTEN_SEND_COOKIE_REQUEST) {
             WPACKET wpkt;
-            unsigned int version;
```
Comment (t8m) [inline ssl/d1_lib.c:668]: Please move the declaration of the wpkt inside the construct_ functions unless there is a good reason to keep it here.

File: ssl/d1_lib.c [hunk 4]
```diff
@@ -485,9 +645,16 @@ int DTLSv1_listen(SSL *ssl, BIO_ADDR *client)
         OPENSSL_free(buf);
         return -1;
     }
+    cookie = OPENSSL_malloc(65535); // Todo: magic number, haven't found a definition in the source code yet.
+    if (cookie == NULL) {
+        OPENSSL_free(buf);
+        OPENSSL_free(wbuf);
+        return -1;
+    }
```
Comment (t8m) [inline ssl/d1_lib.c:653]: Why is this change needed, especially why the fairly large allocation size?

**Review by t8m — CHANGES_REQUESTED** [review]

File: ssl/d1_lib.c (hunk 4 above)
Comment (fwh-dc) [reply ssl/d1_lib.c:653]: My rationale was that since RFC 8446 section 4.2.2 defines the cookie as:

>      struct {
>          opaque cookie<1..2^16-1>;
>       } Cookie;

This code should support the larger size, but maybe it was never the intention for this code? Any case I've reverted the code to the old functionality.

Comment (t8m) [reply ssl/d1_lib.c]: I am actually not sure about that. Perhaps @mattcaswell has an opinion?

**Comment by mattcaswell** [comment]
Looks like this needs a rebase

**Review by andrewkdinh — CHANGES_REQUESTED** [review]
Please add tests for this as well, particularly for testing `ClientHello`

**Comment by fwh-dc** [comment]
@andrewkdinh Thanks for your comments. This is still a draft and is missing a rebase etc. I will reply to your comments when I have time to work on this part of the DTLS 1.3 feature.
